sol/main.c

```diff
@@ -90,63 +97,107 @@ insert_new_priority_req(struct req_queue *req_queue, struct priority_req *pr,
+ * Note, that list is expected to be not empty.
+ */
+static inline struct rte_mbuf *
+list_first_entry_m(struct list_head *ptr) {
+	return rte_priv_to_mbuf(ptr->next);
+}
+
+/*
+ * Get the last rte_mbuf element from a list.
+ * Note, that list is expected to be not empty.
+ */
+static inline struct rte_mbuf *
+list_last_entry_m(struct list_head *ptr) {
+	return rte_priv_to_mbuf(ptr->prev);
+}
+
+/* Get the prev rte_mbuf element in list. */
+static inline struct rte_mbuf *
+list_prev_entry_m(struct rte_mbuf *pos) {
+	return list_to_mbuf(mbuf_to_list(pos)->prev);
+}
+
+/* Get the next rte_mbuf element in list. */
+static inline struct rte_mbuf *
+list_next_entry_m(struct rte_mbuf *pos) {
+	return list_to_mbuf(mbuf_to_list(pos)->next);
+}
+
 static void
-drop_lowest_priority_pkt(struct sol_config *sol_conf,
-	struct req_queue *req_queue)
+drop_lowest_priority_pkt(struct req_queue *req_queue)
 {
-	struct priority_req *lowest_pr = list_last_entry(&req_queue->head,
-		struct priority_req, list);
-	struct priority_req *next_lowest_pr;
+	struct rte_mbuf *lowest_pr = list_last_entry_m(&req_queue->head);
+	struct rte_mbuf *next_lowest_pr;
 
 	RTE_VERIFY(req_queue->len > 0);
 
 	if (unlikely(req_queue->len == 1)) {
-		req_queue->priorities[lowest_pr->priority] = NULL;
+		req_queue->priorities[lowest_pr->udata64] = NULL;
 		req_queue->highest_priority = 0;
 		req_queue->lowest_priority = GK_MAX_REQ_PRIORITY;
 		goto drop;
 	}
 
-	next_lowest_pr = list_prev_entry(lowest_pr, list);
+	next_lowest_pr = list_prev_entry_m(lowest_pr);
 
 	/* The lowest priority packet was the only one of that priority. */
-	if (lowest_pr->priority != next_lowest_pr->priority) {
-		req_queue->priorities[lowest_pr->priority] = NULL;
-		req_queue->lowest_priority = next_lowest_pr->priority;
+	if (lowest_pr->udata64 != next_lowest_pr->udata64) {
+		req_queue->priorities[lowest_pr->udata64] = NULL;
+		req_queue->lowest_priority = next_lowest_pr->udata64;
 		goto drop;
 	}
 
-	req_queue->priorities[lowest_pr->priority] = next_lowest_pr;
+	req_queue->priorities[lowest_pr->udata64] = next_lowest_pr;
 
 drop:
-	list_del(&lowest_pr->list);
-	rte_pktmbuf_free(lowest_pr->pkt);
-	mb_free_entry(&sol_conf->mb, lowest_pr);
+	list_del(mbuf_to_list(lowest_pr));
+	rte_pktmbuf_free(lowest_pr);
 	req_queue->len--;
 }
 
 static void
-enqueue_req(struct sol_config *sol_conf, struct priority_req *req)
+enqueue_req(struct sol_config *sol_conf, struct rte_mbuf *req)
 {
 	struct req_queue *req_queue = &sol_conf->req_queue;
-	uint8_t priority = req->priority;
+	uint8_t priority = req->udata64;
 
 	if (req_queue->len >= sol_conf->pri_req_max_len) {
 		/* New packet is lowest priority, so drop it. */
 		if (req_queue->lowest_priority >= priority) {
-			rte_pktmbuf_free(req->pkt);
-			mb_free_entry(&sol_conf->mb, req);
+			rte_pktmbuf_free(req);
 			return;
 		}
-		drop_lowest_priority_pkt(sol_conf, req_queue);
+		drop_lowest_priority_pkt(req_queue);
 	}
 
+	INIT_LIST_HEAD(mbuf_to_list(req));
```

Comment:
Neither `list_add()`, or `insert_new_priority_req()` below require `req` to be initialized, so don't initialize it here. But add a comment to `enqueue_req()` and `insert_new_priority_req()` saying that they do not require that the list field of a request to be initialized.